File: Neocom/Neocom/NCMainMenuHeaderViewController.m
//
//  NCMainMenuHeaderViewController.m
//  Neocom
//
//  Created by Artem Shimanski on 14.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

#import "NCMainMenuHeaderViewController.h"
#import "NCAccount.h"
#import "NCDataManager.h"

@interface NCMainMenuHeaderViewController ()

@end

@implementation NCMainMenuHeaderViewController

- (void)viewDidLoad {
    [super viewDidLoad];
	NCAccount* account = NCAccount.currentAccount;
	self.characterNameLabel.text = @" ";
	self.corporationLabel.text = @" ";
	self.allianceLabel.text = @" ";
	self.characterImageView.image = nil;
	self.corporationImageView.image = nil;
	self.allianceImageView.image = nil;
	
	self.corporationLabel.superview.hidden = YES;
	self.allianceLabel.superview.hidden = YES;
	
	NCDataManager* dataManager = [NCDataManager defaultManager];
	if (account.eveAPIKey.corporate) {

	}
	else {
		[dataManager characterInfoForAccount:account cachePolicy:NSURLRequestUseProtocolCachePolicy completionHandler:^(EVECharacterInfo *result, NSError *error, NSManagedObjectID *cacheRecordID) {
			if (error) {
			}
			else {
				self.characterNameLabel.text = result.characterName;
				self.corporationLabel.text = result.corporation;
				self.allianceLabel.text = result.alliance;
				self.corporationLabel.superview.hidden = result.corporationID == 0;
				self.allianceLabel.superview.hidden = result.allianceID == 0;

				if (self.characterImageView)
					[dataManager imageWithCharacterID:result.characterID preferredSize:CGSizeMake(128, 128) scale:UIScreen.mainScreen.scale cachePolicy:NSURLRequestUseProtocolCachePolicy completionBlock:^(UIImage *image, NSError *error) {
						self.characterImageView.image = image;
					}];
				if (self.corporationImageView && result.corporationID)
					[dataManager imageWithCorporationID:result.corporationID preferredSize:CGSizeMake(32, 32) scale:UIScreen.mainScreen.scale cachePolicy:NSURLRequestUseProtocolCachePolicy completionBlock:^(UIImage *image, NSError *error) {
						self.corporationImageView.image = image;
					}];
				if (self.allianceImageView && result.allianceID)
					[dataManager imageWithAllianceID:result.allianceID preferredSize:CGSizeMake(32, 32) scale:UIScreen.mainScreen.scale cachePolicy:NSURLRequestUseProtocolCachePolicy completionBlock:^(UIImage *image, NSError *error) {
						self.allianceImageView.image = image;
					}];
			}
		}];
	}
	
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
	if ([segue.identifier isEqualToString:@"NCAccountsViewController"]) {
		segue.destinationViewController.transitioningDelegate = (id) self.parentViewController;
	}
	[super prepareForSegue:segue sender:sender];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)onLogout:(id)sender {
	NCAccount.currentAccount = nil;
}

@end

File: Neocom/Neocom/NCMainMenuHeaderViewController.swift
//
//  NCMainMenuHeaderViewController.swift
//  Neocom
//
//  Created by Artem Shimanski on 04.12.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import UIKit
import EVEAPI

class NCMainMenuHeaderViewController: UIViewController {
	@IBOutlet weak var characterNameLabel: UILabel?
	@IBOutlet weak var characterImageView: UIImageView?
	@IBOutlet weak var corporationLabel: UILabel?
	@IBOutlet weak var allianceLabel: UILabel?
	@IBOutlet weak var corporationImageView: UIImageView?
	@IBOutlet weak var allianceImageView: UIImageView?
	@IBOutlet weak var heightConstraint: NSLayoutConstraint?
	
	override func viewDidLoad() {
		super.viewDidLoad()
		characterNameLabel?.text = " "
		corporationLabel?.text = " "
		allianceLabel?.text = " "
		characterImageView?.image = nil
		corporationImageView?.image = nil
		allianceImageView?.image = nil
		allianceLabel?.superview?.isHidden = true
		
		if let account = NCAccount.current {
			let progressHandler = NCProgressHandler(view: view, totalUnitCount: 6)
			let dataManager = NCDataManager(account: account)
			let characterID = account.characterID
			OperationQueue(qos: .utility).async {
				progressHandler.progress.perform { dataManager.image(characterID: characterID, dimension: 128) }
					.then(on: .main) { result in
						self.characterImageView?.image = result.value
				}

				let character = try progressHandler.progress.perform { dataManager.character() }
					.then(on: .main) { result -> ESI.Character.Information in
						guard let value = result.value else {throw NCDataManagerError.noCacheData}
						self.characterNameLabel?.text = value.name
						return value
					}.catch(on: .main) { error in
						self.characterNameLabel?.text = error.localizedDescription
					}.get()
				
				progressHandler.progress.perform { dataManager.image(corporationID: Int64(character.corporationID), dimension: 32) }
					.then(on: .main) { result in
						self.corporationImageView?.image = result.value
				}

				let corporation = try progressHandler.progress.perform { dataManager.corporation(corporationID: Int64(character.corporationID)) }
					.then(on: .main) { result -> ESI.Corporation.Information in
						guard let corporation = result.value else {throw NCDataManagerError.noCacheData}
						self.corporationLabel?.text = corporation.name
						return corporation
					}.catch(on: .main) { error in
						self.corporationLabel?.text = error.localizedDescription
					}.get()
				

				if let allianceID = corporation.allianceID {
					progressHandler.progress.perform { dataManager.image(allianceID: Int64(allianceID), dimension: 32) }
						.then(on: .main) { result in
							self.allianceImageView?.image = result.value
					}

					progressHandler.progress.perform { dataManager.alliance(allianceID: Int64(allianceID)) }
						.then(on: .main) { result in
							guard let alliance = result.value else {throw NCDataManagerError.noCacheData}
							self.allianceLabel?.text = alliance.name
							self.allianceLabel?.superview?.isHidden = false
						}.catch(on: .main) { error in
							self.allianceLabel?.text = error.localizedDescription
							self.allianceLabel?.superview?.isHidden = true
						}.wait()
				}
				else {
					DispatchQueue.main.async {
						self.allianceLabel?.superview?.isHidden = true
					}
				}
			}.finally {
				progressHandler.finish()
			}
		}
	}
	
	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
		if segue.identifier == "NCAccountsViewController" {
			segue.destination.transitioningDelegate = parent as? UIViewControllerTransitioningDelegate
		}
	}

	@IBAction func onLogout(_ sender: Any) {
		NCAccount.current = nil
	}
	
	@IBAction func onAddAccount(_ sender: Any) {
		ESI.performAuthorization(from: self)

//		let url = OAuth2.authURL(clientID: ESClientID, callbackURL: ESCallbackURL, scope: ESI.Scope.default, state: "esi")
//		if #available(iOS 10.0, *) {
//			UIApplication.shared.open(url, options: [:], completionHandler: nil)
//		} else {
//			UIApplication.shared.openURL(url)
//		}
		
	}
}
